File: 2021/Day3/3sum/3sum.py
class Solution:#By CyFun
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        counter = Counter(nums)
        nums, tripluts = list(counter.keys()), set()
        if counter[0] >= 3:
            tripluts.add((0, 0, 0))
        plus, minus = [n for n in nums if n > 0], [
            n for n in nums if n < 0]
        for a in minus:
            for b in plus:
                c = -(a + b)
                if c in counter and ((c != a and c != b) or counter[c] > 1):
                    tripluts.add(tuple(sorted([a, b, c])))
        return tripluts

File: 2021/Day3/3sum/3sum.cpp
class Solution //By CyFun
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        int n = nums.size();
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        if (n < 3)
            return {};

        for (int i = 0; i < n - 2; i++)
        {

            if (i == 0 || nums[i] != nums[i - 1])
            {
                int j = i + 1, k = n - 1;
                while (j < k)
                {
                    int sum = nums[i] + nums[j] + nums[k];
                    if (sum == 0)
                    {
                        ans.push_back({nums[i], nums[j], nums[k]});
                        while (j < k && nums[j] == nums[j + 1])
                            j++;
                        while (j < k && nums[k] == nums[k - 1])
                            k--;
                        j++;
                        k--;
                    }
                    else if (sum > 0)
                    {
                        k--;
                    }
                    else
                        j++;
                }
            }
        }
        return ans;
    }
};